src/test/java/org/broadinstitute/consent/http/service/DatasetRegistrationServiceTest.java

```diff
@@ -0,0 +1,531 @@
+
+    @Mock
+    private GCSService gcsService;
+
+    @Before
+    public void setUp() {
+        MockitoAnnotations.openMocks(this);
+    }
+
+    private void initService() {
+        datasetRegistrationService = new DatasetRegistrationService(datasetDAO, dacDAO, datasetServiceDAO, gcsService);
+    }
+
+
+    // captor: allows you to inspect the arguments sent to a function.
+    @Captor
+    ArgumentCaptor<List<DatasetServiceDAO.DatasetInsert>> datasetInsertCaptor;
+
+    // ------------------------ test multiple dataset insert ----------------------------------- //
+    @Test
+    public void testInsertCompleteDatasetRegistration() throws Exception {
+        User user = mock();
+        DatasetRegistrationSchemaV1 schema = createRandomCompleteDatasetRegistration(user);
+
+        spy(gcsService);
+
+        initService();
+
+        Map<String, FormDataBodyPart> files = Map.of("alternativeDataSharingPlan", createFormDataBodyPart(), "consentGroups[0].nihInstitutionalCertificationFile", createFormDataBodyPart(), "otherUnused", createFormDataBodyPart());
+        when(gcsService.storeDocument(any(), any(), any())).thenReturn(BlobId.of("asdf", "hjkl"), BlobId.of("qwer", "tyuio"));
+        when(dacDAO.findById(any())).thenReturn(new Dac());
+
+        datasetRegistrationService.createDatasetsFromRegistration(schema, user, files);
+
+        verify(datasetServiceDAO).insertDatasets(datasetInsertCaptor.capture());
+
+        // only two files are stored; extra "unused" file not used
+        verify(gcsService, times(2)).storeDocument(any(), any(), any());
+
+        List<DatasetServiceDAO.DatasetInsert> inserts = datasetInsertCaptor.getValue();
+
+        assertEquals(1, inserts.size());
+
+        assertEquals(schema.getDataAccessCommitteeId(), inserts.get(0).dacId());
+        assertEquals(schema.getConsentGroups().get(0).getConsentGroupName(), inserts.get(0).name());
+        assertDataUse(schema.getConsentGroups().get(0), inserts.get(0).dataUse());
+        assertEquals(user.getUserId(), inserts.get(0).userId());
+
+        assertEquals(2, inserts.get(0).files().size());
+        assertEquals(FileCategory.ALTERNATIVE_DATA_SHARING_PLAN, inserts.get(0).files().get(0).getCategory());
+        assertEquals(files.get("alternativeDataSharingPlan").getContentDisposition().getFileName(), inserts.get(0).files().get(0).getFileName());
+        assertEquals(BlobId.of("asdf", "hjkl"), inserts.get(0).files().get(0).getBlobId());
+
+        assertEquals(FileCategory.NIH_INSTITUTIONAL_CERTIFICATION, inserts.get(0).files().get(1).getCategory());
+        assertEquals(files.get("consentGroups[0].nihInstitutionalCertificationFile").getContentDisposition().getFileName(), inserts.get(0).files().get(1).getFileName());
+        assertEquals(BlobId.of("qwer", "tyuio"), inserts.get(0).files().get(1).getBlobId());
+
+
+        List<DatasetProperty> props = inserts.get(0).props();
+        assertContainsDatasetProperty(props, "piName", schema.getPiName());
```

Comment:
How can we ensure that we haven't missed a property in this list?

Reply:
Hmm, the complication is that not all fields of the dataset registration are turned into properties - only a subset - so we cannot use reflection or something like that to check they all exist. Unless we basically copied over the extractors - which would make the tests depend and use the potentially flawed implementation of the registration system, I am not sure there is a great way to ensure that. I am with you that it's not ideal though. I could break it up visually more, but I'm not sure that solves the problem.

Reply:
Perhaps worth adding a TODO comment here.